File: BennuLib/IO/FpgSpriteAssetDecoder.vb
Option Infer On

Imports PkSprite.Core

Namespace Bennu.IO
    Public Class FpgSpriteAssetDecoder
        Inherits NativeDecoder(Of SpriteAsset)

        Public Overrides ReadOnly Property MaxSupportedVersion As Integer = &H0 ' Current readable Fpg version

        Protected Overrides ReadOnly Property KnownFileExtensions As String() = {"fpg"}

        Protected Overrides ReadOnly Property KnownFileIds As String() = {"fpg", "f16", "f32"}

        Protected Overrides Function ReadNativeFormat(magic As Magic, reader As NativeFormatReader) As SpriteAsset
            Dim depth = magic.Depth
            If depth = 8 Then
                Dim pal = Palette.Create(VGAtoColors(reader.ReadPalette))
                reader.ReadUnusedPaletteGamma()
            End If

            Dim fpg As New SpriteAsset

            Try
                Do
                    Dim code = reader.ReadInt32
                    Dim maplen = reader.ReadInt32
                    Dim description = reader.ReadDescription
                    Dim name = reader.ReadChars(12).ToString()
                    Dim width = reader.ReadInt32
                    Dim height = reader.ReadInt32

                    Dim mapDataLength = width * height * (depth \ 8)

                    Dim numberOfPivotPoints = reader.ReadPivotPointsNumber()
                    Dim pivotPoints() = reader.ReadPivotPoints(numberOfPivotPoints)

                    ' Serves as checksum for FPGs created with non-standard tools such
                    ' as FPG Edit
                    If mapDataLength + 64 + numberOfPivotPoints * 4 <> maplen Then
                        Exit Do
                    End If


                    Dim graphicData = reader.ReadBytes(mapDataLength)
                    Dim pixels = CreatePixelBuffer(magic.Depth, graphicData)

                    Dim map = Sprite.Create(width, height, pixels)
                    map.Description = description

                    fpg.Update(code, map)
                Loop
            Catch exception As System.IO.EndOfStreamException
                ' Do nothing (the reading of the map
            End Try

            Return fpg
        End Function
    End Class
End Namespace

File: BennuLib.Sharp/IO/FpgSpriteAssetDecoder.cs
using static BennuLib.IO.NativeFormat;

namespace BennuLib.IO
{
	public class FpgSpriteAssetDecoder : NativeDecoder<SpriteAsset>
	{

        public override int MaxSupportedVersion { get; } = 0x00;

        protected override string[] KnownFileExtensions { get; } = { "fpg" };

        protected override string[] KnownFileMagics { get; } = { "f16", "f32", "fpg", "f01" };

		protected override SpriteAsset ReadBody(Header header, NativeFormatReader reader)
		{
            SpriteAsset fpg;

            if (header.Depth == 8)
            {
                Palette palette = Palette.Create(VGAtoColors(reader.ReadPalette()));
                reader.ReadUnusedPaletteGamma();
                fpg = SpriteAsset.Create(palette);
            }
            else
            {
                fpg = SpriteAsset.Create((DepthMode)header.Depth);
            }

			try
            {
				do
                {
					var code = reader.ReadInt32();
					var maplen = reader.ReadInt32();
					var description = reader.ReadAsciiZ(32);
					var name = reader.ReadAsciiZ(12);
					var width = reader.ReadInt32();
					var height = reader.ReadInt32();

					var mapDataLength = width * height * (header.Depth / 8);

					var numberOfPivotPoints = reader.ReadPivotPointsNumberLong();
					var pivotPoints = reader.ReadPivotPoints(numberOfPivotPoints);

					// Some tools such as FPG Edit are non conformant with the standard
                    // FPG files and will add data at the end. 
					if (mapDataLength + 64 + numberOfPivotPoints * 4 != maplen) {
                        // It can be that many tools generate this field with wrong
                        // information. Check SmartFpgEditor output!
                        // break; 
                        // TODO: Consider if for example, we shall generate some 
                        // kind of event
					}

					var graphicData = reader.ReadBytes(mapDataLength);
					var pixels = CreatePixelBuffer(header.Depth, graphicData);

					var map = Sprite.Create(width, height, pixels);
					map.Description = description;
                    foreach (var point in pivotPoints)
                    {
                        map.DefinePivotPoint(point.Id, point.X, point.Y);
                    }

					fpg.Update(code, map);

				} while (true);

			}
            catch (System.IO.EndOfStreamException)
            {
				// Do nothing. The file is consumed until it is not possible to 
                // read any more data.
			}

			return fpg;
		}
	}
}
